Separate gammasimp from combsimp & Improve combsimp()

ylemkimon opened the pull request:
* Removed original `as_gamma`, `as_factorial`, `as_binomial` flags
* Removed rewriting binomial coefficient to `_rf((n - k + 1).expand(), k.expand())/_rf(1, k.expand()))`. Fixes #11548 and fixes #8156
* Added binomial coefficient simplification with `binomial(n, n-r) = binomial(n, r)`. Fixes #6658 
* Expand gamma functions with rational number argument. Fixes #6153 and fixes #10101
* Added rewrite as binomial coefficient: finds `(a+b)!/a!b!` and replaces with binomial coefficient. Fixes #7135 and #6341

---

* Separated gamma function simplification(`gammasimp`) from `combsimp`. Fixes #9785, see #9785 and #13132 for related dicussions and table below.
* `combsimp` does not apply gamma function simplifications that may make an integer argument non-integer. Fixes #6341
* Using `combsimp` to simplify expression involving gamma functions or combinatorial functions with non-integer argument is passed to `gammasimp` automatically

Arguments | combsimp() | gammasimp()
----------- | ------------ | ------------
integer | factorial, binomial | gamma 
non-integer | gamma | gamma

ylemkimon committed:
Improve combsimp

* Added as_comb as option
* Removed as_gamma and as_factorial
* Removed binomial rewrite to simplified form
* Binomial simplification
* Simplify gamma only when as_comb is False
* Gamma expansion
* Rewrite as binomial coefficient

asmeurer commented:
Not sure about the deprecation.

asmeurer commented on sympy/simplify/gammasimp.py:
```diff
@@ -0,0 +1,527 @@
+    where possible and expands gammas with rational argument.
+
+    If as_comb is True, it rewrites gammas as factorials and converts
+    (a+b)!/a!b! into binomials.
```
I wonder if this flag is really needed here. The former can be done with `rewrite(gamma)` or `replace`, and the latter should be done by combsimp. I get that you want to keep the logic together for now, but perhaps that should be hidden from the user.

ylemkimon replied:
@asmeurer I agree. How about making _gamma_as_comb a public function with maybe force flag to determine to check for integer arguments?

asmeurer replied:
Would it do any more than expr.rewrite(factorial) (or maybe `combsimp(expr.rewrite(factorial))`)?

ylemkimon commented:
@asmeurer I think there are several options of result of `combsimp(functions with non-integer args)`:
1. gamma w/ DeprecationWarning, possibly ValueError in the future
2. gamma
3. factorial, binomial
4. factorial, binomial if force=True, else gamma w/ DeprecationWarning, possibly ValueError in the future
4. factorial, binomial if force=True, else gamma

asmeurer commented:
I don't like the options that give warnings or errors. SymPy otherwise allows non-integer arguments to combinatorial functions. Apparently `factorial(0.5)` even evaluates to `gamma(1.5)`. 

2 or 5 seem nice, but I don't feel strongly about it.

ylemkimon commented:
@jksuom What do you think?

jksuom commented:
I would have preferred that `factorial` raise an error on manifestly non-integer arguments like 0.5. However, it currently returns None and then mpmath evaluates it by means of `gamma` if the argument is a `Float`.

Maybe a similar automatic (silent) coercion `factorial -> gamma` could be applied also in `combsimp`. The arguments would be checked (the second arguments of `rf` and `ff`), and if they cannot be integers, then the expression would be passed over to `gammasimp`. Then `compsimp` could deal with combinatorial functions only (with no keywords like `force`).

ylemkimon commented:
* Removed as_comb option from gammasimp, as same result can be acheived by combsimp
* Instead of deprecation, combsimp passes expression with non-integer arguments to gammasimp automatically

ylemkimon commented:
@asmeurer @jksuom Thank you.

---
* Removed as_comb option from gammasimp, as same result can be acheived by combsimp
* Instead of deprecation, combsimp passes expression with non-integer arguments to gammasimp automatically

asmeurer commented:
Apparently SymPy has evaluated factorial as gamma for a long time. I bisected it to 7e7427adbbc62c655b2fd17243c5fb83c1ad485c. I agree with @jksuom that it would be better if it didn't. We could consider changing it.

ylemkimon commented:
@asmeurer Doesn't other CASes and calculators do this? Also other combinatorial functions like binomial doesn't have analytical counterpart, and it may be inconsistent that only factorial raises error.

asmeurer commented:
That's a good question. What do other systems do?

I'll also point out that it could also remain unevaluated instead of giving an error.

ylemkimon commented:
@asmeurer 
* **[Python math](https://docs.python.org/2/library/math.html#math.factorial)** : 
> Raises ValueError if x is not integral or is negative.
* **[mpmath](http://mpmath.readthedocs.io/en/latest/functions/gamma.html#factorials)** : 
> more generally the factorial is defined for real or complex x by x!=Γ(x+1)
* **[Maxima](http://maxima.sourceforge.net/docs/manual/maxima_10.html#factorial)** :
> For a real or complex number in float or bigfloat precision x, x! simplifies to the value of gamma (x+1). For x equal to n/2 where n is an odd integer, x! simplifies to a rational factor times sqrt(pi) (since gamma (1/2) is equal to sqrt(pi)). 
* **[Maple](https://www.maplesoft.com/support/help/maple/view.aspx?path=factorial)** :
> If m is a (real or complex) floating-point number, Maple returns the generalized factorial function result calculated using GAMMA(m+1).
* **[Wolfram(Mathematica)]** : 
> For non‐integer n, the numerical value of n! is given by Gamma[1+n]. For integers and half integers, Factorial automatically evaluates to exact values.

ylemkimon commented:
@asmeurer @jksuom Please review, thank you.

jksuom commented on sympy/simplify/combsimp.py:
```diff
@@ -64,450 +47,82 @@ def combsimp(expr):
+                    n_args.extend([factor]*exp)
+                    if isinstance(factor, factorial):
+                        n_fact_args.extend([factor.args[0]]*exp)
+                except TypeError:
```
Under which conditions does this occur?

ylemkimon replied:
@jkusom If exp is non-integer(not int castable).

jksuom commented on sympy/simplify/combsimp.py:
```diff
@@ -64,450 +47,82 @@ def combsimp(expr):
+                try:
+                    d_args.extend([factor]*exp)
+                except TypeError:
+                    n_args.append(factor**exp)
```
Why `n_args` instead of `d_args`?

ylemkimon replied:
@ jksuom Typo probably from copy and paste, will fix it.

jksuom commented:
This seems to be working well. Some imports are in functions. Could they be moved to the top of the file?

jksuom commented on sympy/simplify/combsimp.py:
```diff
@@ -64,450 +47,82 @@ def combsimp(expr):
+                    d_args[i] = None
+                    d_args[j] = None
+                    hit = True
                     break
```
Could there be another loop where the roles of `n` and `d` have been interchanged?

ylemkimon commented:
@jksuom 
I wonder whether there are better factorials to binomial algorithm or heuristics.

jksuom commented:
>  I wonder there are better factorials to binomial algorithm/heuristics.

I don't know of any. I think that the idea of `_gamma_as_comb` is good. The implementation could probably be extended to handle possible binomials also in the denominator.

ylemkimon commented:
@jksuom Fixed 😄 

---
* Refactored factorials to binomial coefficient algorithm in _gamma_as_comb().
* Added detection of inverse of binomial coefficient
* Moved gammasimp imports from function to top.

ylemkimon commented:
I'm not sure why following test fails on Travis:
```
________________________________________________________________________________
_____________ sympy/concrete/tests/test_gosper.py:test_gosper_sum ______________
  File "/home/travis/virtualenv/python2.7.13/lib/python2.7/site-packages/sympy/concrete/tests/test_gosper.py", line 48, in test_gosper_sum
    + 1/(gamma(a)*gamma(b))
AssertionError
```
It successfully runs on local machine(Windows).

ylemkimon committed:
Separate gammasimp from combsimp

Separated gamma function simplification(gammasimp) from combsimp. Also
separated _gamma_as_comb from combsimp and refactored its algorithm,
adding detection of inverse of binomial coefficient. See
sympy/sympy#9785 and sympy/sympy#13132.

jksuom commented on sympy/simplify/combsimp.py:
```diff
@@ -64,450 +47,70 @@ def combsimp(expr):
+                        hit = True
+
+                        nd_fact_args[1 - m].remove(sum)
+                        del nd_fact_args[m][j]
```
Deleting items from a list while looping may lead to errors. Setting them to None was probably safer.

ylemkimon replied:
@jksuom I thought it's safe since it is guaranteed that `j > i`, broken after `del`, and does not use list iterator. Or does it depends on the implementation of Python?

jksuom replied:
I'm not sure. Maybe it will always work as `len(nd_fact_args[m])` is recomputed for each loop index `i`.

jksuom commented on sympy/simplify/combsimp.py:
```diff
@@ -64,450 +47,70 @@ def combsimp(expr):
-                    new.append(g1)
-            # /!\ updating IN PLACE
-            gammas[:] = new
+    from .simplify import bottom_up
```
Could this be at the top as well?

ylemkimon replied:
@jksuom As `sympy.simplify.simplify` imports `combsimp`, it creates circular import.

ylemkimon replied:
I think we should move `bottom_up` from `simplify.py` to `traversaltools.py`, but it's another issue.

jksuom commented:
This looks good to me. Thanks.
